recipes/yasm/all/conanfile.py

```diff
@@ -14,87 +18,90 @@ class YASMConan(ConanFile):
 
     def package_id(self):
         del self.info.settings.compiler
 
     def source(self):
-        tools.get(**self.conan_data["sources"][self.version][0],
-                  destination=self._source_subfolder, strip_root=True)
-        tools.download(**self.conan_data["sources"][self.version][1],
-                       filename=os.path.join(self._source_subfolder, "YASM-VERSION-GEN.bat"))
+        get(self, **self.conan_data["sources"][self.version][0],
+                  destination=self.source_folder, strip_root=True)
+        download(self, **self.conan_data["sources"][self.version][1],
+                       filename=os.path.join(self.source_folder, "YASM-VERSION-GEN.bat"))
 
     @property
     def _msvc_subfolder(self):
-        return os.path.join(self._source_subfolder, "Mkfiles", "vc10")
+        return os.path.join(self.source_folder, "Mkfiles", "vc10")
 
-    def _build_vs(self):
-        with tools.chdir(self._msvc_subfolder):
-            msbuild = MSBuild(self)
-            if self.settings.arch == "x86":
-                msbuild.build_env.link_flags.append("/MACHINE:X86")
-            elif self.settings.arch == "x86_64":
-                msbuild.build_env.link_flags.append("/SAFESEH:NO /MACHINE:X64")
-            build_type = "Debug" if self.settings.build_type == "Debug" else "Release"
-            msbuild.build(project_file="yasm.sln", build_type=build_type, upgrade_project=False,
-                          targets=["yasm"], platforms={"x86": "Win32"}, force_vcvars=True)
-
-    def _configure_autotools(self):
-        if self._autotools:
-            return self._autotools
-        self._autotools = AutoToolsBuildEnvironment(self, win_bash=tools.os_info.is_windows)
-        yes_no = lambda v: "yes" if v else "no"
-        conf_args = [
-            "--enable-debug={}".format(yes_no(self.settings.build_type == "Debug")),
+    def _generate_vs(self):
+        tc = MSBuildToolchain(self)
+        tc.generate()
+
+    def _generate_autotools(self):
+        tc = AutotoolsToolchain(self)
+        enable_debug = "yes" if self.settings.build_type == "Debug" else "no"
+        tc.configure_args.extend([
+            f"--enable-debug={enable_debug}",
             "--disable-rpath",
             "--disable-nls",
-        ]
-        self._autotools.configure(args=conf_args, configure_dir=self._source_subfolder)
-        return self._autotools
+        ])
+        tc.generate()
+
+    def generate(self):
+        if is_msvc(self):
+            self._generate_vs()
+        else:
+            self._generate_autotools()
 
     def build(self):
-        if self.settings.compiler == "Visual Studio":
-            self._build_vs()
+        apply_conandata_patches(self)
+        if is_msvc(self):
+            msbuild = MSBuild(self)
+            msbuild.build_type = "Debug" if self.settings.build_type == "Debug" else "Release"
+            msbuild.platform = "Win32" if self.settings.arch == "x86" else msbuild.platform
```

Comment:
I am surprised these are not automatic 🤔 https://github.com/conan-io/conan/blob/dd2965c92da7bab30f3b483b54be24a9504ba4bb/conan/tools/microsoft/msbuild.py#L22\

If it's not working might be a bug

```suggestion
```